### ISSUE TITLE
Automatic differentiation for tuning uncertain molecular data and/or computing exact Jacobian spectra.

### ISSUE THREAD

**Issue by gully** [radis/radis#153]
### 🎯 Describe what you'd like

The physical properties of each spectral line can be thought of as tunable parameters--- adjustable within the uncertain of their published values, with some molecular data more certain than others'.  When a synthetic spectrum is compared to high fidelity data, the values of these parameters can be fine-adjusted via [backpropagation](https://en.wikipedia.org/wiki/Backpropagation) and [stochastic gradient descent](https://en.wikipedia.org/wiki/Stochastic_gradient_descent), algorithms that scale to very large number of parameters, possibly millions or billions of parameters.  This design is now routinely applied to training the voluminous weights of Neural Networks.  

### 💡 Possible implementations

Several frameworks from the Neural Network community enable the automatic computation of derivatives, such as [PyTorch](https://pytorch.org/) and [jax](https://github.com/google/jax).  I have made prototypes to illustrate the technical feasibility of this approach for astronomical spectroscopy:

In [this notebook](https://github.com/gully/tgif) we show some toy spectrum example to understand the key concepts and provide a proof-of-concept for autodiff.

In [this notebook](https://github.com/BrownDwarf/fiatlux/blob/master/notebooks/09_atmospheric_retrieval.ipynb) I demo how to compute exact Jacobian spectra (e.g. _d Spectrum / d Pressure_) using Jax.  We also compute the Hessian Matrix in a data-model comparison to obtain uncertainties in a stratified P-T profile for Earth's atmosphere.  The demo uses HAPI just for I/O, and reimplements the computation of opacities with `jax.numpy` so that it is autodiff aware.  Jax also supports JIT and GPUs natively.

I'm currently working on a demo with PyTorch to make the optimization of a large number of parameters more streamlined, also using HAPI for the I/O.  

In practice, supporting autodiff would represent a major design change for radis, and I don't expect it to be a near-term goal, and is probably an altogether separate project.  However, I'm posting it here to encourage some experimentation now in preparation for a next generation code.  

### The main barriers I anticipate are: 

GPU Memory demands.  The frameworks need to store---in memory---the network graph of every computation that has occurred involving each free parameter.  Depending on how the code is structured, this graph can become very large and so GPU memory becomes the bottleneck (as it already is in radis).  [Keops](https://www.kernel-operations.io/keops/index.html) is one way to get more mileage out of a single GPU.  

### Side benefits

**Out of the box support for most relevant functions.**  codes like PyTorch and Jax support differentiating through FFT's out of the box, so at least one ostensible risk is already retired: line profile convolutions will work as they currently do.

**New ways to generate code from the Partition Function**  This @mattjj [YouTube video](https://youtu.be/ALMKFS_QFm4?t=283) hints at how Thermodynamic quantities like Gibbs Free Energy could be computed automatically with autodiff.  One virtue of such strategies may reside in the automatic code generation aspects and gains in numerical stability by differentiation through `logsumexp`.  Minimizing these quantities to find equilibrium compositions, for example, is also made effortless through higher order derivatives in jax.  It's even possible that multilayer atmospheric codes could be made to satisfy differential equations that are themselves implemented as gradients of thermodynamic quantities, analogous to the notion of [implicit layers](https://implicit-layers-tutorial.org/) in neural networks.

**Comment by erwanp** [comment]
Thank you for the details here, and for checking that FFTs were implemented ! 

As you say, i also anticipate large changes in RADIS. Our best call is to do it the other way round, i.e., implement the LDM method of RADIS #37  (paper just out !) into [fiatlux](https://github.com/BrownDwarf/fiatlux/blob/master/notebooks/09_atmospheric_retrieval.ipynb) or directly in [blase](https://github.com/gully/blase). 

Let's discuss further on Slack [#autodiff](https://radis-radiation.slack.com/archives/C01G3117DJS) .

**Comment by erwanp** [comment]
Automatic Differentiation was implemented in https://github.com/HajimeKawahara/exojax  by @HajimeKawahara
